Adding Localization

nyamsprod opened the pull request:
A DataLocalizer class is added. This class enables localizing the country full name using:
- PHP intl extension.
- the data from `ISO3166::KEY_ALPHA3`.

The public API is as follow:

``` php
public DataLocalizer::__construct($key = 'name', $locale = '')
public DataLocalizer::localize(iterable $iterable): Generator
public DataLocalizer::__invoke(iterable $iterable): Generator
// __invoke is an alias of the localize method
```

A simple usage will be:

``` php
use League\ISO3166\ISO3166;
use League\ISO3166\DataLocalizer;

$localizer = new DataLocalizer('local_name', 'fr');
$collection = new ISO3166;
$res1 = $localizer($collection);
$res2 = $localizer->localize($collection->listBy(ISO3166::KEY_ALPHA2);
$res3 = $localizer->__invoke([$collection->getByAlpha3('SEN')]);
```

The result is a generator you can iterate over where each entry has a new key `local_name` with the country name being localized as its value.
- If the alpha3 code is unknown: the new key will return the alpha3 code
- If the locale is unknown: the new key will return the localized name according to PHP `local_get_default`

A possible BC break is that this feature relies on PHP intl extension which is AFAIK bundle by default since PHP5.5.

This feature may resolve #11

nyamsprod committed:
Adding Localization

Adding the DataLocalizer class.
This class expect an iterable and will return a generator with the localized name of
the country using PHP intl extension.

alcohol commented:
As much as I appreciate the work, I think I would rather create an additional package called `iso3166-intl`, which both adds this support and also introduces the extra dependency on `ext-intl` (assuming the intl extension is the preferred solution, I am not sure of that yet).

Ocramius commented on src/DataLocalizer.php:
```diff
@@ -0,0 +1,141 @@
+use InvalidArgumentException;
+use Traversable;
+
+class DataLocalizer
```
Missing interface?

nyamsprod replied:
Not at all adding an interface would effectively make localization intl independent. I'll look into it

Ocramius commented on src/DataLocalizer.php:
```diff
@@ -0,0 +1,141 @@
+     *
+     * @return Generator
+     */
+    public function localize($iterable)
```
Any reason for an alias API?

nyamsprod replied:
Well the only public method is `DataLocalizer::localize` adding `DataLocalizer::__invoke` make easier to use IMHO. anyway if an interface is created the `__invoke` will not be part of the interface

Ocramius replied:
I see it the other way around: the only public method is `__invoke`, while `localize` is a useless verb-alike name required because the class is given a noun. If you rename the class to `LocalizeData` you are much better off anyway

nyamsprod committed:
DataLocalizer is now an Interface

DataLocalizer is made an Interface and IntlLocalizer implements DataLocalizer
using the php-intl extension
